### PR TITLE
Add state to distributed composable API

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87954
* #87664
* __->__ #87838

### CODE REVIEW

File: test/distributed/_composable/test_contract.py [hunk 1]
```diff
@@ -93,6 +93,30 @@ def wrap_module(module: nn.Module) -> nn.Module:
...
+    @skipIfTorchDynamo("Dynamo does not yet capture module hooks")
+    def test_state(self):
+        def check_and_update_state_hook(
+            module: nn.Module, inp: Tuple[torch.Tensor]
+        ) -> Tuple[torch.Tensor]:
+            self.assertEqual(api.state(module).dummy_state, 7)
+            api.state(module).dummy_state = 8
+            return inp
+
+        # FIXME: circular reference looks a bit weird. Shall we make .state a
```
Comment (mrshenli) [inline test/distributed/_composable/test_contract.py:105]: Any thoughts on this?

File: torch/distributed/_composable/contract.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    # implement operator < to avoid breaking dir()
+    def __lt__(self, other: Any) -> bool:
+        return True if isinstance(other, str) else id(self) < id(other)
+
+
+class _State:
+    pass
+
+
+state_key = _StateKey()
```
Comment (yhcharles) [inline torch/distributed/_composable/contract.py:22]: How about `STATE_KEY`, like a constant?

File: torch/distributed/_composable/contract.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        >>> my_feature(model.l2)
+        >>> model(torch.randn(2, 10)).sum().backward()
+    """
+
+    def get_all_state(module: nn.Module) -> Dict[Callable, Dict]:
+        d = module.__dict__.setdefault(state_key, {})
+        assert isinstance(d, dict), "Distributed composable API states corrupted"
+        return d
+
+    def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
```
Comment (yhcharles) [inline torch/distributed/_composable/contract.py:66]: Do we want to make `module` -> `*module`, like in the design doc?

File: torch/distributed/_composable/contract.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        assert func not in all_state, (
+            "Each distinct composable distributed API can only be applied to a "
+            f"module once. {func} has already been applied to the following "
+            f"module.\n{module}"
+        )
+        all_state.setdefault(func, _State())
+
+        orig_named_params = OrderedDict(module.named_parameters())
+        orig_named_buffers = OrderedDict(
+            module.named_buffers(remove_duplicate=False)
```
Comment (yhcharles) [inline torch/distributed/_composable/contract.py:78]: Seems `module.named_buffers()` has no arg of `remove_duplicate`:
https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_buffers

File: torch/distributed/_composable/contract.py (hunk 3 above)
Comment (mrshenli) [reply torch/distributed/_composable/contract.py:66]: good point. Will update. Is it OK to update that in the follow up PR?

File: torch/distributed/_composable/contract.py (hunk 4 above)
Comment (mrshenli) [reply torch/distributed/_composable/contract.py:78]: The code has the arg:

https://github.com/pytorch/pytorch/blob/1e1b04512879d6166dc1f5adff482723e2d0da9e/torch/nn/modules/module.py#L2011

Let me update `nn.Module`'s doc as well

File: torch/distributed/_composable/contract.py (hunk 3 above)
Comment (yhcharles) [reply torch/distributed/_composable/contract.py:66]: Sure, not a blocker for me

File: torch/distributed/_composable/contract.py (hunk 4 above)
Comment (yhcharles) [reply torch/distributed/_composable/contract.py:78]: Oh, I see, it's added recently: https://github.com/pytorch/pytorch/pull/84984 :)

**Review by yhcharles — APPROVED** [review]

File: torch/distributed/_composable/contract.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    def get_all_state(module: nn.Module) -> Dict[Callable, Dict]:
+        d = module.__dict__.setdefault(state_key, {})
+        assert isinstance(d, dict), "Distributed composable API states corrupted"
+        return d
+
+    def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
+        # install states specific to the wrapped ``func``
+        all_state: Dict[Callable, dict] = get_all_state(module)
+        assert func not in all_state, (
```
Comment (yhcharles) [inline torch/distributed/_composable/contract.py:69]: An open question we can resolve later: how to make sure some APIs are mutual exclusive, for example shard_embedding/replicate, while some others are not, for example checkpoint/fsdp.

Comment (mrshenli) [reply torch/distributed/_composable/contract.py:69]: one option could be letting the contract adding info to the state, and the contract can help checking whether 1) same API is called twice, 2) whether there are conflicting APIs. 

But might need a way to allow new APIs to declare conflicts.

**Pull request closed without merging**